RHIDP-4780: Document per-ConfigMap/Secret configuration of mountPath (#842)
redhat-developer/red-hat-developers-documentation-rhdh: modules/configuring/proc-mounting-additional-files-in-your-custom-configuration-using-rhdh-operator.adoc

* If `key` and `mountPath` are not specified: Each key or value is mounted as a `filename` or content with a `subPath`.
* If `key` is specified with or without `mountPath`: The specified key or value is mounted with a `subPath`.
* If only `mountPath` is specified: A directory containing all the keys or values is mounted without a `subPath`.

[NOTE]
====
* {ocp-short} does not automatically update a volume mounted with subPath. By default, the {product-very-short} operator monitors these Config Maps or Secrets and refreshes the {product-very-short} Pod when changes occur.
* For security purposes, {product} restricts the Operator Service Account's read access to Secrets. As a result, mounting files from Secrets without specifying both mountPath and key is not supported.
====

.Prerequisites
* You have developer permissions to access the {ocp-short} cluster containing your {product-short} instance using the {openshift-cli}.
* link:{installing-on-ocp-book-url}[Your {ocp-short} administrator has installed the {product} Operator in {ocp-short}].

.Procedure

> . In {ocp-short}, author your ConfigMap or Secret custom resource in a YAML file. For more information, see link:{installing-on-ocp-book-url}#provisioning-your-custom-configuration[].

Reviewer: True that this link does not seem to be rendering properly, but _For more information, see [link]_ is the proper format, so leave that part as written. Let's investigate the link issue.

> . In {ocp-short}, author your ConfigMap or Secret custom resource in a YAML file. For more information, see link:{installing-on-ocp-book-url}#provisioning-your-custom-configuration[].

Author: I see! Thanks so much for the help with the Cross references, Lindsey. I'll keep that in mind.
Yes, I also found "author" a bit odd. I went with your suggestion and I found a mention of this in another doc:
https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.4/html/telemetry_data_collection/assembly-rhdh-telemetry#proc-enabling-telemetry-using-operator_title-telemetry

It says, "Create xyz ConfigMap with the following YAML code"...

+
.Minimal `{my-app-config-config-map}` custom resource example
====
[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: {my-product-cr-name}
data:
  file11.txt: |
    My file11 content
  file 12.txt: |
    My file12 content
----
====
+

> .Minimal `{my-product-secrets}` custom resource example

Author: This section is under Chapter 1.2. So, this is Chapter 1.2.1. So, there are a few more examples in the section before this one. That's why it's showing as Example 7 and Example 8 instead of Example 1 and Example 2.

====
[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: Secret
metadata:
  name: {my-product-secrets}
StringData:
  secret11.txt: |
    secret-content
----
====

> . Declare the ConfigMap or Secret file in your `{product-custom-resource-type}` custom resource.

Reviewer: Is _Declare_ the right word? Maybe. But the linter for 8th grade language might flag it and suggest that a simpler word can be used.

> . Declare the ConfigMap or Secret file in your `{product-custom-resource-type}` custom resource.

Author: I don't think so either. I rewrote it as per the text I found here:
https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.4/html/telemetry_data_collection/assembly-rhdh-telemetry#proc-enabling-telemetry-using-operator_title-telemetry

But I am a little doubtful about the accuracy of 'configMaps name' and 'secrets name'. Also, I feel the line - "Set the value of the 'configMaps name' or 'secrets name' to the name of the ConfigMap or Secret file......." could be slightly confusing. Would it be a good idea to split this into two separate lines, by saying, "Set the value of the 'configMaps name' to the name of the ConfigMap file or the 'secrets name' to the name of the Secret file in your `{product-custom-resource-type}` CR."  ?

> . Declare the ConfigMap or Secret file in your `{product-custom-resource-type}` custom resource.

Reviewer: Yeah, I agree that the clarity of that step can be improved with an edit like the one you suggested - good catch!

[id="mounting-additional-files-in-your-custom-configuration-using-rhdh-operator"]
= Mounting additional files in your custom configuration using the {product} operator

> You can use the {product-short} Operator to mount extra files, such as ConfigMaps and Secrets, to the container in a preferred location.

Reviewer 2: ```suggestion
You can use the {product-short} Operator to mount extra files, such as config maps and secrets, to the container in a preferred location.
```

The `mountPath` field specifies the location where a ConfigMap or Secret is mounted. The behavior of the mount, whether it includes or excludes a `subPath`, depends on the specification of the `key` or `mountPath` fields.

* If `key` and `mountPath` are not specified: Each key or value is mounted as a `filename` or content with a `subPath`.
* If `key` is specified with or without `mountPath`: The specified key or value is mounted with a `subPath`.
* If only `mountPath` is specified: A directory containing all the keys or values is mounted without a `subPath`.

[NOTE]
====

> * {ocp-short} does not automatically update a volume mounted with subPath. By default, the {product-very-short} operator monitors these ConfigMaps or Secrets and refreshes the {product-very-short} Pod when changes occur.

Reviewer 3: ```suggestion
* {ocp-short} does not automatically update a volume mounted with `subPath`. By default, the {product-very-short} operator monitors these ConfigMaps or Secrets and refreshes the {product-very-short} Pod when changes occur.
```

* For security purposes, {product} does not give the Operator Service Account read access to Secrets. As a result, mounting files from Secrets without specifying both mountPath and key is not supported.
====

.Prerequisites
* You have developer permissions to access the {ocp-short} cluster containing your {product-short} instance using the {openshift-cli}.
* link:{installing-on-ocp-book-url}[Your {ocp-short} administrator has installed the {product} Operator in {ocp-short}].

.Procedure

> . In {ocp-short}, create your config map or secret with the following YAML codes:

Reviewer: ```suggestion
. In {ocp-short}, create your ConfigMap or Secret with the following YAML codes:
```

> . In {ocp-short}, create your ConfigMap or Secret custom resource (CR) with the following YAML codes:

Reviewer 2: Config maps and secrets are not custom resources, but standard Kubernetes objects.

Additionally, when referring to API object terms in general usage, use lowercase and separate multi-word API objects. 
* when referring to the resource name, write: config map, secret, custom resource
* When referring to the API object, write:  `ConfigMap` object, `Secret` object, `CustomResource` object. See https://redhat-documentation.github.io/supplementary-style-guide/#config-map and https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#general-references

```suggestion
. In {ocp-short}, create your config map or secret with the following YAML:
```

Additionally

+

> .Minimal `{my-extra-file-configmap}` config map example

Reviewer: ```suggestion
.Minimal `{my-extra-file-configmap}` ConfigMap example
```

> .Minimal `{my-app-config-config-map}` CR example

Reviewer 2: ```suggestion
.Minimal `{my-app-config-config-map}` config map example
```

====
[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: ConfigMap
metadata:

>   name: {my-product-cr-name}

Reviewer 4: name: {my-product-configmap}

Author: There's a slight change in the attributes. I've changed it to {my-app-config-config-map} which displays "my-rhdh-app-config".

data:
  file11.txt: |
    My file11 content
  file 12.txt: |
    My file12 content
----
====
+

> .Minimal `{my-product-secrets}` secret example

Reviewer: ```suggestion
.Minimal `{my-product-secrets}` Secret example
```

> .Minimal `{my-product-secrets}` CR example

Reviewer 2: ```suggestion
.Minimal `{my-product-secrets}` secret example
```

====
[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: Secret
metadata:
  name: {my-product-secrets}
StringData:
  secret11.txt: |
    secret-content
----
====
For more information, see xref:provisioning-your-custom-configuration[Provisioning and using your custom {product} configuration].

> . Set the value of the `configMaps name` or 'secrets name' to the name of the config map or secret file in your `{product-custom-resource-type}` CR. For example:

Reviewer: ```suggestion
. Set the value of the `configMaps name` or `secrets name` to the name of the config map or secret file in your `{product-custom-resource-type}` CR. For example:
```
@pabel-rh I fixed this formatting, but I like your suggestion to rewrite this step for better clarity

> ====

Reviewer 5: ```suggestion
+
====
```
Indent example.

> . Set the value of the `configMaps name` to the name of the config map or `secrets name` to the name of the secret file in your `{product-custom-resource-type}` CR. For example:

Reviewer: This is must clearer - thank you! The only comment I have is that Kubernetes formats these objects as ConfigMap and Secret (capitalized that way) so we should do the same in our docs. Please forgive me for not noticing that earlier! 
```suggestion
. Set the value of the `configMaps name` to the name of the ConfigMap or `secrets name` to the name of the Secret in your `{product-custom-resource-type}` CR. For example:
```

Author: That's okay! Sure, I'll keep that in mind.

[source,yaml,subs="+attributes,+quotes"]
----
spec:
  application:
    extraFiles:
      mountPath: /my/path
      configMaps:

>         - name: {my-app-config-config-map}

Reviewer 4: name: {my-product-configmap}

Reviewer 2: {my-app-config-config-map} is the correct attribute. We initially chose not to use {my-product-config-map} attribute name because we provision usually at least 3 different config maps: dynamic-plugins, app-config, rbac-policies, and that would have raised clarity issues.

Reviewer 2: But in this example, we are speaking about a random custom map. We should not use an attribute that is already used somewhere else.

Author: I've discussed with @hmanwani-rh . We are going with my-project-configmap. I've created a new attribute for extra files. It's called "my-extra-file-configmap". I'm open to suggestions. :) 
Heena suggests that it's better we say "my-project-configmap" instead of "my-product-configmap" to differentiate between the user's project with the RHDH application.

          key: file12.txt
          mountPath: /my/my-rhdh-config-map/path
      secrets:

>         - name: my-rhdh-secret

Reviewer 4: name: {my-product-secrets}
